root/static/scripts/common/components/Autocomplete2/recentItems.js

```diff
@@ -0,0 +1,272 @@
+  const storedMap = _getStoredMap();
+  const storedValueList = storedMap[key];
+
+  if (storedValueList != null && Array.isArray(storedValueList)) {
+    for (const value of storedValueList) {
+      /*
+       * We previously stored recent entities in their entirety, as
+       * JSON objects, but that caused issues when the expected
+       * format changed from what was stored. This is why we now
+       * only deal with MBIDs and fetch those from the web service
+       * instead.
+       */
+      switch (typeof value) {
+        case 'object': {
+          if (value) {
+            const id = _getGidOrId(value);
+            if (id != null) {
+              ids.add(id);
+            }
+          }
+          break;
+        }
+        case 'string': {
+          ids.add(value);
+          break;
+        }
+      }
+      if (ids.size >= MAX_RECENT_ENTITIES) {
+        break;
+      }
+    }
+  }
+
+  return ids;
+}
+
+function _setRecentEntityIds(
+  key: string,
+  ids: Set<string> | null,
+): void {
+  const storedMap = _getStoredMap();
+  storedMap[key] = ids ? [...ids] : [];
+
+  localStorage(
+    'recentAutocompleteEntities',
+    JSON.stringify(storedMap),
+  );
+}
+
+export function clearRecentItems(
+  key: string,
+): void {
+  _setRecentEntityIds(key, null);
+  _recentItemsCache.set(key, []);
+}
+
+const _recentItemsCache =
+  new Map<string, $ReadOnlyArray<OptionItem<EntityItem>>>();
+
+export function getRecentItems<+T: EntityItem>(
+  key: string,
+): $ReadOnlyArray<OptionItem<T>> {
+  // $FlowIgnore[incompatible-return]
+  return _recentItemsCache.get(key) ?? [];
+}
+
+export async function getOrFetchRecentItems<+T: EntityItem>(
+  entityType: string,
+  key?: string = entityType,
+): Promise<$ReadOnlyArray<OptionItem<T>>> {
+  const ids = _getRecentEntityIds(key);
+  const cachedList = [...getRecentItems(key)];
+
+  _recentItemsCache.set(key, cachedList);
+
+  for (const item of cachedList) {
+    const id = _getGidOrId(item.entity);
+    if (id != null) {
+      ids.delete(id);
+    }
+  }
+
+  if (ids.size) {
+    // Convert ids to an array since we delete in the loop.
+    for (const id of Array.from(ids)) {
+      const entity: ?EntityItem = linkedEntities[entityType]?.[id];
+      if (entity) {
+        cachedList.push({
+          // $FlowIgnore[incompatible-call]
+          entity,
+          id: String(entity.id) + '-recent',
+          name: entity.name,
+          type: 'option',
+        });
+        ids.delete(id);
+      }
+    }
+  }
+
+  if (ids.size) {
+    // $FlowIgnore[incompatible-return]
+    return fetch(
+      '/ws/js/entities/' +
+      entityType + '/' +
+      Array.from(ids.values()).join('+'),
+    ).then((resp) => {
+      if (!resp.ok) {
+        return null;
+      }
+      return resp.json();
+    }).then((data: WsJsEntitiesDataT | null) => {
+      if (!data) {
+        return cachedList;
+      }
+
+      const results = data.results;
+      invariant(cachedList);
```

Comment:
It’s the first usage ever of this function you just introduced. Please explain (in commit message?) why is that needed here.

Reply:
That's explained on https://github.com/metabrainz/musicbrainz-server/pull/2124/commits/84363005bf7013adef02f257538365a28f6fe6f9 (not saying it's not confusing, though...)

Reply:
It's not actually needed here, so I removed it.  (Must've been left from a previous version of the code.)  I added a small example to the commit message where I introduce the function.

Reply:
Thanks!